### PR TITLE
[contractstaking] improve robustness

### PR DESCRIPTION
**Pull request by envestcc**
# Description
improve code robustness according to new comments in #3888 

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: blockindex/contractstaking/cache.go [hunk 1]
```diff
@@ -223,6 +224,20 @@ func (s *contractStakingCache) LoadFromDB(kvstore db.KVStore) error {
...
 	defer s.mutex.Unlock()
 
+	// load height
+	var height uint64
+	h, err := kvstore.Get(_StakingNS, _stakingHeightKey)
+	if err != nil {
+		if !errors.Is(err, db.ErrNotExist) {
+			return err
+		}
+		height = 0
```
Comment (dustinxie) [inline blockindex/contractstaking/cache.go:234]: add `totalBucketCount = 0` to L247 below, or remove this line, to be consistent

File: blockindex/contractstaking/cache.go [hunk 2]
```diff
@@ -331,21 +358,23 @@ func (s *contractStakingCache) putBucketType(id uint64, bt *BucketType) {
...
 	if _, ok := s.candidateBucketMap[newDelegate]; !ok {
 		s.candidateBucketMap[newDelegate] = make(map[uint64]bool)
 	}
 	s.candidateBucketMap[newDelegate][id] = true
-	// delete old candidate bucket map
-	if oldBi != nil {
-		oldDelegate := oldBi.Delegate.String()
-		if oldDelegate != newDelegate {
-			delete(s.candidateBucketMap[oldDelegate], id)
-		}
```
Comment (dustinxie) [inline blockindex/contractstaking/cache.go:347]: how likely is the case old bucket needs to be deleted? if it's not too much (like < 5-10%), then I think just add 2 lines (to delete empty map here) would suffice:
```
if len(s.candidateBucketMap[oldDelegate]) == 0 {
				delete(s.candidateBucketMap, oldDelegate)
			}
```

File: blockindex/contractstaking/cache.go [hunk 3]
```diff
@@ -23,6 +23,7 @@ type (
 		bucketTypeMap         map[uint64]*BucketType      // map[bucketTypeId]BucketType
 		propertyBucketTypeMap map[int64]map[uint64]uint64 // map[amount][duration]index
 		totalBucketCount      uint64                      // total number of buckets including burned buckets
+		height                uint64                      // current block height
```
Comment (Liuhaai) [inline blockindex/contractstaking/cache.go:26]: add comments for the reason to put `height` in the cache

File: blockindex/contractstaking/cache.go (hunk 2 above)
Comment (Liuhaai) [reply blockindex/contractstaking/cache.go:347]: the efficiency doesn't matter much here, otherwise more `if condition` will harm readability

**Review by Liuhaai — APPROVED** [review]

File: blockindex/contractstaking/cache.go (hunk 2 above)
Comment (dustinxie) [reply blockindex/contractstaking/cache.go:347]: given the same correct logic, i would say we should aim for better code
the `if` here indicates the need to remove old/existing bucket, it's not too excessive `if` i think

**Review by dustinxie — APPROVED** [review]

**Pull request merged**